Makefile

```diff
@@ -63,103 +92,122 @@ push-target:
 	(cd pkg/web && packr2 clean && packr2)
 
 .PHONY: compile-datamon
-## Build datamon executable on local OS
 compile-datamon: export LDFLAGS=${BASE_LD_FLAGS} ${VERSION_LD_FLAGS}
 compile-datamon: build-assets
-	@echo 'building ${YELLOW}datamon${RESET} executable'
-	@echo "${VERSION_LD_FLAGS}"
-	@echo "${LDFLAGS}"
-	go build -o $(TARGET) --ldflags "${LDFLAGS}" ./cmd/datamon
+	@echo '$(GREEN)INFO: building ${YELLOW}datamon${GREEN} executable$(RESET)'
+	@if [[ -z "$(TARGET)" ]] ; then echo "You must specify a TARGET" && exit 1; fi
+	go build -o $(TARGET) -ldflags "${LDFLAGS}" ./cmd/datamon
 
 .PHONY: build-datamon-local
-## DEMO: Build datamon executable in-place for local os.  Goes with internal operationalization demos.
-build-datamon-local: export TARGET=cmd/datamon/datamon
+## BUILD: Build datamon executable for local os
+build-datamon-local: export TARGET=out/datamon
 build-datamon-local:
+	@mkdir -p out
 	$(MAKE) compile-datamon
 
+
 .PHONY: build-datamon-mac
-## Build datamon executable for mac os x (on mac os x)
+## BUILD: Build local datamon executable for mac os x in ./out directory (on mac os x)
 build-datamon-mac: export TARGET=out/datamon.mac
 build-datamon-mac:
 	@mkdir -p out
 	$(MAKE) compile-datamon
 
-.PHONY: build-all
-## Build all the containers
-build-all: clean build-datamon build-migrate
-
-.PHONY: build-migrate
-## Build migrate tool
-build-migrate: export BUILD_TARGET=reg.onec.co/migrate
-build-migrate: export DOCKERFILE=migrate.Dockerfile
-build-migrate: export BUILD_ARGS=--pull
-build-migrate: export TAGS=$(USER_TAG) $(BRANCH_TAG)
-build-migrate:
-	$(MAKE) build-target
+.PHONY: cross-compile-binaries
+## BUILD: Build all binaries with cross compilation for linux and darwin platforms
+cross-compile-binaries: export TARGET=out/bin
+cross-compile-binaries: export OS=linux darwin
+cross-compile-binaries: export ARCH=amd64
+cross-compile-binaries: export LDFLAGS=${BASE_LD_FLAGS} ${VERSION_LD_FLAGS}
+cross-compile-binaries: build-assets
+	@TOOLS=github.com/mitchellh/gox $(MAKE) ensure-gotools
+	@mkdir -p ${TARGET}
+
+	# Ref: https://github.com/mitchellh/gox/issues/55 for CGO_ENABLED=0
+	CGO_ENABLED=0 gox -os "${OS}" -arch "${ARCH}" -tags netgo -output "${TARGET}/{{.Dir}}_{{.OS}}_{{.Arch}}" -ldflags "${LDFLAGS}" ./...
+
+	@if [[ -n "${UPX_FOR_OSX}" ]]; then \
+		echo "INFO: compressing all binaries" && \
+		upx ${TARGET}/* ; \
+	elif [[ -n "${UPX_VERSION}" ]] ; then \
+	  echo "INFO: compressing linux binaries only. Install upx >=3.96 to compress darwin binaries" && \
+		upx ${TARGET}/*_linux_* ; \
+	else \
+	  echo "WARN:: upx not available in this environment. Binaries not compressed." ; \
+	fi
+
+.PHONY: build-and-push-datamon-builder
+## BUILD: Build new base builder image for CI jobs
+build-and-push-datamon-builder:
+	$(BUILDER) -t $(REPOSITORY)/datamon-builder:${BASEVERSION} - < dockerfiles/builder.Dockerfile
+	docker tag $(REPOSITORY)/datamon-builder:${BASEVERSION} $(REPOSITORY)/datamon-builder:latest
```

Comment:
How does this work with parallel builds? Who gets the latest tag? Since this is based on branch, should the tag be the branch or hash of the commit? For the release, we can have a check for the tag attached to the commit?

Reply:
this step was not intended to be run during PR runs, but by cronjob only and from master only.
the content of this image does not depend on the repo